### PR TITLE
Add oneDNN Graph fusion pass in Inductor CPP backend

### LINKED ISSUE #105582: RFC: Integrating oneDNN Graph Compiler into Inductor C++/OpenMP Backend for Enhanced Graph Fusion and Performance
### 🚀 The feature, motivation and pitch

Integrating oneDNN Graph Compiler into Inductor C++ Backend enables enhanced pattern fusion and performance for CPU. 
design

### Motivation
Recent developments on the Inductor C++ backend have demonstrated promising performance on DL inference workloads with CPU, thanks to optimizations like Conv/GEMM + post-op fusions and vectorization (see [this](https://dev-discuss.pytorch.org/t/Inductor-update-4-cpu-backend-started-to-show-promising-performance-boost/874) and [this](https://dev-discuss.pytorch.org/t/torchinductor-update-5-cpu-backend-backend-performance-update-and-deep-dive-on-key-optimizations/1117)). 

[oneDNN Graph API](https://spec.oneapi.io/onednn-graph/latest/introduction.html) (codename LLGA) extents oneDNN with a high-level graph API. It goes beyond Conv/GEMM post-op fusions and supports [aggressive fusion patterns](http://oneapi-src.github.io/oneDNN/dev_guide_graph_fusion_patterns.html#aggressive-fusion-patterns) such as MultiheadAttention, MLP blocks, and more (with its [graph compiler backend](http://oneapi-src.github.io/oneDNN/dev_guide_graph_compiler.html)). Other features include  [low precision](http://oneapi-src.github.io/oneDNN/dev_guide_graph_low_precision.html). Since PyTorch 1.12, this API has been added in [TorchScript JIT fuser path](https://pytorch.org/tutorials/recipes/recipes/tuning_guide.html#use-onednn-graph-with-torchscript-for-inference) showing promising performance [#49444](https://github.com/pytorch/pytorch/issues/49444).  

Integrating the oneDNN Graph Compiler with Inductor C++ backend offers further performance enhancements. Additionally, adopting the oneDNN Graph API simplifies design and development.

### Plan
Our long-term goal is to use oneDNN Graph fusion by default, replacing post-op fusions. Starting as an experimental feature, we have implemented a `onednn_graph_fusion` pass into the Inductor post-graph passes, enabled by an inductor.cpp config option. We hope it will eventually become the default option after validation of no performance regression from Inductor C++ backend.

We start with CPU inference for float32 data type. In the future, we will add support for other PyTorch 2 features, like training, quantization, dynamic shape, BF16, etc.. 

#### Inductor Post-grad pass
We introduce the `onednn_graph_fusion` pass directly takes the FX graph from AOTAutograd as input. The FX graph is used to construct an LLGA Graph by lowering aten/prims IR to LLGA IR and by lowering `FakeTensor` to LLGA `LogicalTensor`. LLGA identifies fusion opportunities in the Graph and returns a list of LLGA `partition`s, each represents a set of fused operations. 

To enable the desired fusion, we rewrite the FX graph by fusing nodes within each LLGA partition into a call_function node. The target of the call_function node points to the corresponding LLGA kernel, which represents the compiled partitions. Other aten ops in the FX graph are executed by the Inductor C++ backend.

#### Other graph rewrites
Some graph rewrites are implemented to match current oneDNN fusion patterns and simplify the LLGA graph. For example, the LLGA op `MatMul` supports any number of dimensions while the `aten.bmm` op only supports 3D inputs, so we rewrite the FX graph to remove `ND -> 3D` and `3D -> ND` transformations before and after a call to `aten.bmm`. These graph rewrites will eventually be supported in the oneDNN Graph code, but currently remain part of our implementation. 

#### Additional context
To implement the backend in Python, we plan to add python binding for oneDNN graph API inside `jit/codegen/onednn` temporarily.

### User interface:
```
torch.compile(options={"cpp.onednn_graph": True})
```

### Preliminary performance
OneDNN Graph provides significant speedups on models that contain advanced fusion patterns such as transformer models, with potentially larger gains as the batch size increases. As the fuison-pattern coverage of oneDNN Graph increases, we expect speedups in more and more models. Currently, the oneDNN-graph pass has an accuracy rate of 56/68 models in the torchbench suite.

The following performance results are from a 32-core, single node test on Sapphire Rapids 

| Benchmark Set (Geometric Mean) | Inductor Speedup | Inductor + oneDNN Graph Speedup |
| -- | -- | -- |
| Torchbench | 1.29x | 1.13x |
| Torchbench  (BS=32) | 1.35x | 1.28x |
| Hugging Face in Torchbench (10 models) | 1.21x | 1.38x |

| Model | Batch Size | Inductor Speedup | Inductor + oneDNN Graph Speedup |
| -- | -- | -- | -- |
| Resnet50 | 32 | 1.72x | 1.81x |
| hf_GPT2_large | 1 | 1.29x | 1.40x |
| hf_GPT2_large | 4 | 1.14x | 1.29x |
| hf_GPT2_large | 32 | 1.22x | 1.47x |
| hf_T5_large | 1 | 1.22x | 2.03x |

<details><summary>Performance Details</summary>
<p>

Model | Batch Size | Accuracy (oneDNN) |   | oneDNN 3.2 | oneDNN+GC 3.2 | Inductor
-- | -- | -- | -- | -- | -- | --
BERT_pytorch | 2 | pass |   | 0.81 | 0.82 | 1.50
Background_Matting | 1 | pass |   | 0.90 | 0.90 | 1.13
LearningToPaint | 96 | pass |   | 1.21 | 1.22 | 1.27
Super_SloMo | 6 | pass |   | 1.11 | 1.11 | 1.17
alexnet | 128 | pass |   | 1.31 | 0.94 | 1.36
attention_is_all_you_need_pytorch | 32 | fail_accuracy |   | 0.82 | 0.16 | 0.91
basic_gnn_edgecnn | 1 | pass |   | 1.60 | 1.58 | 1.66
basic_gnn_gcn | 1 | pass |   | 0.37 | 0.37 | 0.49
basic_gnn_gin | 1 | pass |   | 0.55 | 0.56 | 0.53
basic_gnn_sage | 1 | pass |   | 0.36 | 0.08 | 0.33
cm3leon_generate | 0 | infra_error |   |   |   |  
dcgan | 256 | pass |   | 1.28 | 1.27 | 1.31
densenet121 | 64 | pass |   | 0.96 | 0.97 | 1.66
detectron2_fcos_r_50_fpn | 1 | pass |   | 0.96 | 0.95 | 1.07
dlrm | 2048 | pass |   | 0.90 | 0.96 | 1.17
doctr_det_predictor | 1 | pass |   | 1.34 | 1.35 | 1.40
doctr_reco_predictor | 1 | pass |   | 1.50 | 1.48 | 2.06
drq | 1 | pass |   | 0.61 | 0.09 | 0.99
fastNLP_Bert | 1 | pass |   | 1.24 | 0.08 | 1.27
functorch_dp_cifar10 | 64 | pass |   | 1.15 | 1.02 | 1.03
hf_Albert | 1 | pass |   | 1.19 | 0.06 | 1.29
hf_Bart | 1 | pass |   | 1.01 | 0.03 | 1.16
hf_Bert | 1 | pass |   | 1.27 | 0.02 | 1.25
hf_Bert_large | 1 | pass |   | 2.00 | 0.04 | 1.66
hf_BigBird | 1 | fail_accuracy |   | 0.94 | 0.94 | 1.31
hf_DistilBert | 1 | pass |   | 1.33 | 0.05 | 1.30
hf_GPT2 | 1 | pass |   | 1.12 | 1.10 | 1.01
hf_GPT2_large | 1 | pass_due_to_skip |   | 1.40 | 1.36 | 1.29
hf_Longformer | 0 | fail_to_run |   |   |   |  
hf_Reformer | 1 | fail_accuracy |   | 0.79 | 0.26 | 0.92
hf_T5 | 1 | pass |   | 1.30 | 1.30 | 1.01
hf_T5_base | 1 | pass |   | 1.49 | 1.49 | 1.01
hf_T5_generate | 1 | fail_to_run |   |   |   |  
hf_T5_large | 1 | pass_due_to_skip |   | 2.03 | 2.09 | 1.22
lennard_jones | 1000 | pass |   | 0.65 | 0.66 | 1.26
llama | 32 | pass |   | 0.61 | 0.61 | 0.49
mnasnet1_0 | 32 | pass |   | 1.73 | 1.83 | 2.27
mobilenet_v2 | 16 | pass |   | 1.77 | 1.76 | 2.40
mobilenet_v2_quantized_qat | 0 | fail_to_run |   |   |   |  
mobilenet_v3_large | 32 | pass |   | 2.63 | 2.67 | 3.10
nanogpt_generate | 0 | fail_to_run |   |   |   |  
nvidia_deeprecommender | 256 | fail_accuracy |   | 0.85 | 0.85 | 1.07
opacus_cifar10 | 64 | pass |   | 0.75 | 0.76 | 0.84
phlippe_densenet | 128 | pass |   | 0.71 | 0.72 | 1.80
phlippe_resnet | 128 | pass |   | 1.39 | 1.41 | 1.82
pytorch_CycleGAN_and_pix2pix | 1 | pass |   | 0.90 | 0.91 | 1.18
pytorch_stargan | 16 | pass |   | 0.96 | 0.97 | 0.97
pytorch_unet | 1 | pass |   | 1.13 | 1.14 | 1.06
resnet152 | 32 | pass |   | 1.44 | 1.45 | 1.52
resnet18 | 8 | pass |   | 1.60 | 1.53 | 1.75
resnet50 | 32 | pass |   | 1.81 | 1.82 | 1.72
resnet50_quantized_qat | 0 | fail_to_run |   |   |   |  
resnext50_32x4d | 8 | pass |   | 1.61 | 1.61 | 1.47
sam | 0 | infra_error |   |   |   |  
shufflenet_v2_x1_0 | 64 | pass |   | 1.63 | 1.72 | 2.14
soft_actor_critic | 256 | pass |   | 1.16 | 0.04 | 2.00
speech_transformer | 1 | pass |   | 1.01 | 0.06 | 1.00
squeezenet1_1 | 16 | pass |   | 1.39 | 1.39 | 2.61
timm_efficientnet | 64 | pass |   | 1.40 | 1.67 | 2.23
timm_nfnet | 128 | pass |   | 1.32 | 1.33 | 1.27
timm_regnet | 32 | pass |   | 1.35 | 1.36 | 1.55
timm_resnest | 32 | pass |   | 1.49 | 1.51 | 1.80
timm_vision_transformer | 32 | pass |   | 1.16 | 1.15 | 1.29
timm_vision_transformer_large | 32 | pass_due_to_skip |   | 1.15 | 1.19 | 1.16
timm_vovnet | 32 | pass |   | 1.12 | 1.13 | 1.43
vgg16 | 4 | pass |   | 1.50 | 0.70 | 1.42
vision_maskrcnn | 1 | fail_accuracy |   | 1.26 | 1.04 | 1.29
yolov3 | 8 | pass |   | 1.42 | 1.41 | 1.53
**Geometric Mean Speedup:** |   | 56/68 |   | 1.13 | 0.65 | 1.29

</p>
</details>

### PR DESCRIPTION
**Pull request by wenzhe-nrv**
[RFC: Integrating oneDNN Graph Compiler into Inductor C++/OpenMP Backend for Enhanced Graph Fusion and Performance #105582](https://github.com/pytorch/pytorch/issues/105582)

- [ ] need to update python binding file after #97957

Handling the opaque op is referred to #90356 

cc @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @ngimel @anijain2305

### CODE REVIEW

**Comment by wenzhe-nrv** [comment]
@jgong5 @EikanWang could you please review first? Thanks.

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: torch/_inductor/compile_fx.py [hunk 1]
```diff
@@ -1048,6 +1048,19 @@ def bw_compiler(model: torch.fx.GraphModule, example_inputs):
             boxed_forward_device_index=forward_device,
         )
 
+    if config.cpp.onednn_graph:
+        decompositions = dict(decompositions)
+        from .fx_passes.onednn_graph_fusion import lowerings as onednn_graph_lowerings
+
+        for op in onednn_graph_lowerings.keys() & decompositions.keys():
+            del decompositions[op]
```
[the remark this answers is not in the document]
Comment (frost-intel) [reply torch/_inductor/compile_fx.py:1056]: Yes. We re-apply the Inductor decompositions to any ops that aren't lowered to onednn Graph in `onednn_graph_fusion.reapply_decomps`

File: torch/_inductor/fx_passes/onednn_graph_fusion.py [hunk 2]
```diff
@@ -0,0 +1,821 @@
...
+            output_descs = self.llga_graph.get_compiled_output_descs(
+                compiled_partition, self.partition.get_out_ports()
+            )
+            output_tensors = [
+                torch.empty_strided(out_desc.get_dims(), out_desc.get_strides())
+                for out_desc in output_descs
+            ]
+            return output_tensors[0] if len(output_tensors) == 1 else output_tensors
+
+        if not self.kernel:
```
[the remark this answers is not in the document]
Comment (wenzhe-nrv) [reply torch/_inductor/fx_passes/onednn_graph_fusion.py:144]: @jgong5 are you asking about multiple inference thread to run the model and causing a racing condition? Does launching with `torch.backend.xeon.run_cpu` with one node or one thread use one thread to invoke this function? Any way we could test this?

File: torch/_inductor/exc.py [hunk 3]
```diff
@@ -19,9 +19,9 @@ def _record_missing_op(target):
 class OperatorIssue(RuntimeError):
     @staticmethod
     def operator_str(target, args, kwargs):
-        lines = [f"target: {target}"] + [
-            f"args[{i}]: {arg}" for i, arg in enumerate(args)
-        ]
+        lines = [f"target: {target}"]
+        if not getattr(target, "is_opaque", False):
+            lines += [f"args[{i}]: {arg}" for i, arg in enumerate(args)]
```
[the remark this answers is not in the document]
Comment (frost-intel) [reply torch/_inductor/exc.py:24]: Inductor logs the args for the implicit fallback in `torch/_inductor/graph.py:GraphLowering.call_function`. But for opaque ops each arg is a MultiOutput or InputBuffer object, and when cast to string it recursively prints the parent nodes of each arg.

If a graph is made up of mostly oneDNN partitions, this ends up as exponential recursive calls.

File: test/test_onednn_dynamo.py [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+            return x.clone(memory_format=torch.contiguous_format) @ y
+        with torch.no_grad():
+            x = torch.randn(3, 4, 32, 8)
+            y = torch.randn(3, 4, 8, 16)
+        self._check_fusion(mod, args=(x, y), expected_fusion_names=['bmm_default'])
+        self._compile_and_check_accuracy(mod, args=(x,y))
+
+    def test_no_fail_on_int64_tensor(self):
+        # Testing no crash on int64 tensors
+        # TODO: Does not test LLGA execution, just that graph build doesn't fail
```
[the remark this answers is not in the document]
Comment (wenzhe-nrv) [reply test/test_onednn_dynamo.py]: unsupported datatype will not be lowered to oneDNN Graph. This UT tests no runtime issue when using unsupported datatype with onednn graph. 
We updated the test name `test_no_fail_on_unsupported_dtype`

File: torch/_inductor/fx_passes/onednn_graph_fusion.py [hunk 5]
```diff
@@ -0,0 +1,821 @@
...
+    log.debug("Re-apply Inductor Decomps after fusion for any un-lowered ops")
+    reapply_decomps(gm)
+
+    log.info("Finished compiling graph with oneDNN backend")
+    log.debug("====== Fx Graph after oneDNN compile ======")
+    log.debug(gm.print_readable(print_output=False))
+    return gm
+
+
+llga_intensive_ops = [
```
[the remark this answers is not in the document]
Comment (wenzhe-nrv) [reply torch/_inductor/fx_passes/onednn_graph_fusion.py]: we renamed most of the llga to onednn_graph in onednn_graph_fusion.py file.

File: torch/__init__.py [hunk 6]
```diff
@@ -1512,12 +1512,12 @@ def __eq__(self, other):
     def apply_mode(self, mode: Optional[str]):
         if mode is None or mode == "default":
             pass
-        elif mode in ("reduce-overhead", "max-autotune", "max-autotune-no-cudagraphs"):
+        elif mode in ("reduce-overhead", "max-autotune", "max-autotune-no-cudagraphs", "onednn-graph"):
```
[the remark this answers is not in the document]
Comment (wenzhe-nrv) [reply torch/__init__.py]: Was added for benchmarking convenience. We removed this. And added an env var `TORCHINDUCTOR_ONEDNN_GRAPH`

File: torch/_inductor/__init__.py [hunk 7]
```diff
@@ -83,6 +83,10 @@ def list_mode_options(mode: str = None, dynamic: bool = None) -> Dict[str, Any]:
             "max_autotune": True,
             "triton.cudagraphs": True,
         },
+        # enable onednn graph for c++ backend
+        "onednn-graph": {
+            "cpp.onednn_graph": True,
+        },
```
[the remark this answers is not in the document]
Comment (wenzhe-nrv) [reply torch/_inductor/__init__.py]: removed.

File: test/test_onednn_dynamo.py [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+    allow_manydim_bmm,
+    onednn_graph_fuse_fx,
+    remove_redundant_expand,
+    replace_max_pool_with_indices,
+    replace_t_matmul_to_matmul,
+)
+
+import subprocess
+import os
+os.environ["_DNNL_GRAPH_DISABLE_COMPILER_BACKEND"] = "1"
```
[the remark this answers is not in the document]
Comment (wenzhe-nrv) [reply test/test_onednn_dynamo.py]: removed. Was used to disable graph compiler in UT. But not needed in this PR.

File: test/test_onednn_dynamo.py [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+ONEDNN_BACKEND = "onednn"
```
[the remark this answers is not in the document]
Comment (wenzhe-nrv) [reply test/test_onednn_dynamo.py]: removed.

File: test/test_onednn_dynamo.py [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+    x = x.to(memory_format=torch.channels_last)
+    out = mod(x)
+    """
+        proc = subprocess.Popen(["python", "-c", channels_last_code], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        out = proc.communicate()
+        lines = out[0].decode().split('\n')
+        reorder_lines = [line for line in lines if 'reorder' in line]
+        # Only reorder is tiling conv weights, not input x.
+        self.assertTrue(len(reorder_lines) == 1)
+        self.assertTrue('33x16x3x3' in reorder_lines[0])
```
[the remark this answers is not in the document]
Comment (wenzhe-nrv) [reply test/test_onednn_dynamo.py]: tests removed.

File: test/test_onednn_dynamo.py [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+with torch.no_grad():
+    x = torch.randn(20, 16, 50, 100)
+    x = x.to()
+    out = mod(x)
+    """
+        proc = subprocess.Popen(["python", "-c", channels_last_code], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        out = proc.communicate()
+        lines = out[0].decode().split('\n')
+        conv_ind = [ind for ind, line in enumerate(lines) if 'convolution' in line][0]
+        pooling_ind = [ind for ind, line in enumerate(lines) if 'pooling' in line][0]
```
[the remark this answers is not in the document]
Comment (wenzhe-nrv) [reply test/test_onednn_dynamo.py]: tests removed.

File: test/test_onednn_dynamo.py [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+            compiled_model = torch.compile(mod, options={"cpp.onednn_graph": True})
+            with torch.no_grad():
+                x = torch.randn(4,512,128)
+                compiled_out = compiled_model(x)
+                expected = mod(x)
+            self.assertTrue(torch.allclose(compiled_out[0], expected[0], atol=1e-5))
+            self.assertTrue(torch.allclose(compiled_out[1], expected[1], atol=1e-5))
+            torch._dynamo.reset()
+        os.environ["_DNNL_GRAPH_DISABLE_COMPILER_BACKEND"] = "1"
+
```
[the remark this answers is not in the document]
Comment (wenzhe-nrv) [reply test/test_onednn_dynamo.py]: added UT for MHA and conv resnet block patterns

File: torch/_inductor/fx_passes/onednn_graph_fusion.py [hunk 13]
```diff
@@ -0,0 +1,821 @@
...
+
+
+class LoweringConfig:
+    def __init__(
+        self,
+        llga_op,
+        in_descs_filtering=None,
+        attribute_inputs={},
+        scalar_in_descs=False,
+    ):
```
[the remark this answers is not in the document]
Comment (wenzhe-nrv) [reply torch/_inductor/fx_passes/onednn_graph_fusion.py:643]: added docstring

File: torch/_inductor/fx_passes/onednn_graph_fusion.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jgong5) [reply torch/_inductor/fx_passes/onednn_graph_fusion.py:145]: Yes, race condition from multiple threads in a single process that do inference on the same model.

**Comment by wenzhe-nrv** [comment]
@jiong 
The current location of `llga_graph.py` caused error in some CI cases. 
```
2023-07-28T20:28:51.3084693Z     from torch.csrc.jit.codegen.onednn.llga_graph import LlgaGraph
2023-07-28T20:28:51.3085445Z torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
2023-07-28T20:28:51.3086106Z ModuleNotFoundError: No module named 'torch.csrc'
```
I need to move the file `llga_graph.py` to a new location. This file defines helper python functions for onednn graph python bindings. Where would be a good place to put this file, like somewhere with mkldnn related python files?

**Comment by jgong5** [comment]
> I need to move the file `llga_graph.py` to a new location. This file defines helper python functions for onednn graph python bindings. Where would be a good place to put this file, like somewhere with mkldnn related python files?

Not sure. Need @malfet 's input on this.

File: torch/_inductor/fx_passes/onednn_graph_fusion.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (wenzhe-nrv) [reply torch/_inductor/fx_passes/onednn_graph_fusion.py:145]: @jgong5 We could use `threading.lock()` for the kernel/partition compilation here. What do you think?

[the remark this answers is not in the document]
Comment (wenzhe-nrv) [reply torch/_inductor/fx_passes/onednn_graph_fusion.py:145]: I checked with oneDNN Graph team. partition compile and compiled_partition.execute are multi-thread safe.

[the remark this answers is not in the document]
Comment (jgong5) [reply torch/_inductor/fx_passes/onednn_graph_fusion.py:145]: But it is not from the python side? Suppose there are two threads from python doing inference on this same model when the kernel has not been compiled yet, what would happen?

[the remark this answers is not in the document]
Comment (jgong5) [reply torch/_inductor/fx_passes/onednn_graph_fusion.py:145]: Also, how is dynamic shape handled here?

[the remark this answers is not in the document]
Comment (wenzhe-nrv) [reply torch/_inductor/fx_passes/onednn_graph_fusion.py:145]: Will add a threading Lock to guard the compiling in runtime. 

>how is dynamic shape handled here?

It's handled by Dynamo to recompile when the input shape changes. Thus the kernel will be recreated and compiled.

[the remark this answers is not in the document]
Comment (jgong5) [reply torch/_inductor/fx_passes/onednn_graph_fusion.py]: Is static shape assumed here or some guards are placed to guarantee recompilation?

[the remark this answers is not in the document]
Comment (wenzhe-nrv) [reply torch/_inductor/fx_passes/onednn_graph_fusion.py]: Here assumes static shape. The shape guard for recompilation is in Dynamo. When dynamic shape kernel feature added to oneDNN Graph in future, we would enable the dynamic shape here.

File: torch/csrc/jit/codegen/onednn/llga_graph.py
Comment (jgong5) [inline torch/csrc/jit/codegen/onednn/llga_graph.py:1]: any concern of naming it `onednn_graph.py` and also name the class as `OnednnGraph`? Any reason why we still keep "llga" in the code?

File: torch/_inductor/fx_passes/post_grad.py [hunk 14]
```diff
@@ -64,6 +63,10 @@ def post_grad_passes(gm: torch.fx.GraphModule, is_inference: bool):
         reorder_for_locality(gm.graph)
 
     if config.pattern_matcher:
+        if torch.backends.mkldnn.is_available() and config.cpp.onednn_graph:
+            from .onednn_graph_fusion import onednn_graph_fuse_fx
+
+            onednn_graph_fuse_fx(gm, is_inference)
```
Comment (jgong5) [inline torch/_inductor/fx_passes/post_grad.py:69]: why not put this into `lazy_init`?

**Review by jgong5 — COMMENTED** [review]
Please also address the CI failures.

**Review by jansel — CHANGES_REQUESTED** [review]
This is a pretty massive PR that will be hard to review in a single pass.  Can we break it up into smaller pieces?

I also worry that this isn't reusing Inductor's existing pattern matching infrastructure, so it will cause a pretty big divergence with similar existing optimizations and may be hard to maintain.

File: torch/csrc/jit/codegen/onednn/llga_graph.py
Comment (wenzhe-nrv) [reply torch/csrc/jit/codegen/onednn/llga_graph.py:1]: renamed.

File: torch/_inductor/fx_passes/post_grad.py (hunk 14 above)
Comment (wenzhe-nrv) [reply torch/_inductor/fx_passes/post_grad.py:69]: I tried to put it to `lazy_init`. It requires changing its API by passing `gm` and `is_inference`. Also caused some strange errors in the unittest.  I'm checking.

**Comment by wenzhe-nrv** [comment]
> This is a pretty massive PR that will be hard to review in a single pass. Can we break it up into smaller pieces?

Thanks for your feedback. I understand the concern. We have tried to keep this first PR concise with a limited scope. It's challenging to break it up since it may break the functionalities.  

The core logic can be found in the files added/changed in `torch/_inductor`, especially `torch/_inductor/fx_passes/onednn_graph_fusion.py`. It main implemented a FX graph transformation - building a onednn graph by lowering aten IR to onednn graph IR, then fusing the onednn graph partitions to a call_function node back to the fx graph for execution. 

In addition, we temporarily added python_binding files for onednn graph c++ API to `torch/csrc/jit/codegen/onednn`. These files are being added to oneDNN v3.3 release and will be removed from PyTorch in future when PyTorch uses oneDNN v3.3.

If it would be helpful, I'm available to walk you through the code or address any specific concerns you might have.

> I also worry that this isn't reusing Inductor's existing pattern matching infrastructure, so it will cause a pretty big divergence with similar existing optimizations and may be hard to maintain.

This feature #105582 aims to replace the current mkldnn pattern fusion in long term. In the transition phase, both pattern matching will coexist. 


@jgong5 @ashokei any thought?

**Comment by jansel** [comment]
I think it would be better to just call oneDNN kernels with existing IRs rather than introduce new IRs.

I think we should also start doing something similar to what `select_algorithm.py` does, where it autotunes between different implementations of the same kernel.  Currently, that is just aten vs (many) Triton templates, but there is a PR up to add CUTLASS as third backend.  Using common infra for multiple backends allows that flexibility.

**Comment by ashokei** [comment]
> I think it would be better to just call oneDNN kernels with existing IRs rather than introduce new IRs.

hi @jansel ,  [ oneDNN Graph is little higher abstraction ](https://www.intel.com/content/www/us/en/docs/onednn/developer-guide-reference/2023-2/basic-concepts-002.html) than other template based backends, and allows us to easily add newer/complex subgraph fusions in future. 

as @wenzhe-nrv pointed out, number of lines of code will be drastically reduced if we exclude python bindings (which will be removed in future oneDNN version upgrades).

**Pull request closed without merging**